### ISSUE TITLE
Method selection issue on Oracle JDK 1.8.0_72-ea

### ISSUE THREAD

**Issue by jponge** [eclipse/golo-lang#331]
Using this JDK the `master` branch with #330 from @Artpej fails on a single test: `org.eclipse.golo.compiler.CompileAndRunTest.test_method_invocations`

```
java.lang.AssertionError: 
Expected: is "fixed"
     but: was "variable"
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:8)
	at org.eclipse.golo.compiler.CompileAndRunTest.test_method_invocations(CompileAndRunTest.java:676)
(...)
```

The test case is based around a `String` augmentation and 2 overloaded targets:

```golo
augment java.lang.String {
  function test = |this, a...| -> "variable"
  function test = |this, a| -> "fixed"
}

function test_order_fixed = -> "":test(1)

function test_order_var = -> "":test(1,2)
```

With this particular early access JDK:

```
$ java -version
java version "1.8.0_72-ea"
Java(TM) SE Runtime Environment (build 1.8.0_72-ea-b05)
Java HotSpot(TM) 64-Bit Server VM (build 25.72-b05, mixed mode)
```

then calling `test_order_fixed` selects the variable-arity `test` augmentation rather than the fixed arguments one.

What is also surprising is that `RegularMethodFinder::getCandidates()` properly favours fixed argument targets over variable argument ones:

```java
private List<Method> getCandidates() {
    List<Method> candidates = new LinkedList<>();
    HashSet<Method> methods = new HashSet<>();
    Collections.addAll(methods, receiverClass.getMethods());
    Collections.addAll(methods, receiverClass.getDeclaredMethods());
    for (Method method : methods) {
      if (isCandidateMethod(method) && !method.isVarArgs()) {
        candidates.add(method);
      } else if (isValidPrivateStructAccess(method) && !method.isVarArgs()) {
        candidates.add(method);
      }
    }
    for (Method method : methods) {
      if (isCandidateMethod(method) && method.isVarArgs()) {
        candidates.add(method);
      } else if (isValidPrivateStructAccess(method) && method.isVarArgs()) {
        candidates.add(method);
      }
    }
    return candidates;
  }
```

It is also worth noting that both Travis-CI and the Eclipse Hudson instance haven't spotted the issue, they are running a `u60` JDK 8 at best.

**Comment by jponge** [comment]
Testing on older JDKs still on my machine.

:red_circle: 
```
$ java -version
java version "1.8.0_66-ea"
Java(TM) SE Runtime Environment (build 1.8.0_66-ea-b02)
Java HotSpot(TM) 64-Bit Server VM (build 25.66-b02, mixed mode)
```

:red_circle: 
```
$ java -version
java version "1.8.0_60"
Java(TM) SE Runtime Environment (build 1.8.0_60-b27)
Java HotSpot(TM) 64-Bit Server VM (build 25.60-b23, mixed mode)
```

:red_circle: 
```
$ java -version
java version "1.8.0_40"
Java(TM) SE Runtime Environment (build 1.8.0_40-b25)
Java HotSpot(TM) 64-Bit Server VM (build 25.40-b25, mixed mode)
```

They all fail, which is very strange,

I went through the debugger this afternoon, but I could not find out why this test failed.

**Comment by yloiseau** [comment]
:white_check_mark: “Works on my machine©™” with the fresh new master (commit d85bb24ff735f7)

```
$ java -version
openjdk version "1.8.0_72-internal"
OpenJDK Runtime Environment (build 1.8.0_72-internal-b05)
OpenJDK 64-Bit Server VM (build 25.72-b05, mixed mode)
```

```
$ ./gradlew --daemon clean test
[...]
:test

BUILD SUCCESSFUL
```

**Comment by danielpetisme** [comment]
:red_circle:  "Broke on my machine"
````
$ java version "1.8.0_45"
Java(TM) SE Runtime Environment (build 1.8.0_45-b14)
Java HotSpot(TM) 64-Bit Server VM (build 25.45-b02, mixed mode)
````

**Comment by Artpej** [comment]
fixed! see #333
